aten/src/ATen/native/DispatchStub.h

```diff
@@ -110,10 +113,12 @@ struct CAFFE2_API DispatchStub<rT (*)(Args...), T> {
   std::atomic<FnPtr> cpu_dispatch_ptr;
   FnPtr cuda_dispatch_ptr;
   FnPtr hip_dispatch_ptr;
+  FnPtr fpga_dispatch_ptr;
```

Comment:
why do you need support in DispatchStub?  Are you using TensorIterator?

Reply:
I think this is where I'm really confused. I have replaced the logic in `aten/src/ATen/native/cpu` with `aten/src/ATen/native/fpga`. I am still running the code in `aten/src/ATen/native` to pre-process any optional or keyword arguments because the FPGA can't handle that without reducing performance. 

The modern "FPGA" is always a "CPU + FPGA" linked by OpenCL, as the FPGA itself is not flexible enough to pre-process all of the optional keyword arguments to each kernel. For example, I only implement `mul_out` on the FPGA, not `mul` and `mul_`.

```C++
void mul_kernel(TensorIterator& iter) {
  AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES(iter.dtype(), "mul_out", [&](){
    using vec_t = at::native::ztype<scalar_t>::vec_t;
    fpga_kernel_no_scalar_vec<vec_t>("v__mul", iter);
  });
}
```

`fpga_kernel_no_scalar_vec` streams input `data_ptrs` from the CPU -> FPGA using OpenCL, and then steam output `data_ptrs` from the FPGA -> CPU.

I am using TensorIterator and I maybe don't need it. It does make things much easier though.

Reply:
maybe this becomes partially moot once we implement https://github.com/pytorch/pytorch/issues/29844.

For the device check in `aten/src/ATen/native` i'm not sure how XLA gets around that -- whether we widen the device check or XLA just implements their own thing.  @ailzhang do you know?